src/ibmca_ec.c

```diff
@@ -0,0 +1,918 @@
+                 IBMCA_R_EC_CURVE_DOES_NOT_SUPPORT_SIGNING);
+        return NULL;
+    }
+ #endif
+
+    /* Get curve nid */
+    nid = EC_GROUP_get_curve_name(group);
+    if (nid <= 0) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN_SIG, IBMCA_R_EC_UNSUPPORTED_CURVE);
+        return NULL;
+    }
+
+    /* Create ICA_EC_KEY object */
+    icakey = p_ica_ec_key_new(nid, &privlen);
+    if (icakey == NULL) {
+        /* This curve is not supported by libica. */
+ #ifdef OLDER_OPENSSL
+        ECDSA_METHOD_get_sign(ossl_ecdsa, &sign_setup_sw, &sign_sig_sw);
+ #else
+        EC_KEY_METHOD_get_sign(ossl_ec, &sign_sw, &sign_setup_sw,
+                               &sign_sig_sw);
+ #endif
+        if (sign_sig_sw == NULL || sign_setup_sw == NULL) {
+            IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN_SIG,
+                     IBMCA_R_EC_INTERNAL_ERROR);
+            return NULL;
+        }
+
+        kinv = NULL;
+        r = NULL;
+        ctx = BN_CTX_new();
+        sign_setup_sw(eckey, ctx, &kinv, &r);
+        BN_CTX_free(ctx);
+        return sign_sig_sw(dgst, dgst_len, kinv, r, eckey);
+    }
+
+    /* Get private (D) value from EC_KEY */
+    bn_d = (BIGNUM*)EC_KEY_get0_private_key(eckey);
+    if (bn_d == NULL) {
+        IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN_SIG, IBMCA_R_EC_INTERNAL_ERROR);
+        goto end;
+    }
+
+    /* Format (D) as char array, with leading zeros if necessary */
+    n = privlen - BN_num_bytes(bn_d);
+    for (i = 0; i < n; i++)
+        D[i] = 0x00;
+    BN_bn2bin(bn_d, &(D[n]));
+
+    /* Initialize private ICA_EC_KEY */
+    rc = p_ica_ec_key_init(NULL, NULL, D, icakey);
+    if (rc != 0) {
+        IBMCAerr(IBMCA_F_ICA_EC_KEY_INIT, rc);
+        goto end;
+    }
+
+    /* Call libica signing routine */
+    rc = p_ica_ecdsa_sign(ibmca_handle, icakey, dgst, dgst_len, sigret, sizeof(sigret));
+    if (rc != 0) {
+        /* Possibly no adapter. */
+ #ifdef OLDER_OPENSSL
+        ECDSA_METHOD_get_sign(ossl_ecdsa, &sign_setup_sw, &sign_sig_sw);
+ #else
+        EC_KEY_METHOD_get_sign(ossl_ec, &sign_sw, &sign_setup_sw,
+                               &sign_sig_sw);
+ #endif
+        if (sign_sig_sw == NULL || sign_setup_sw == NULL) {
+            IBMCAerr(IBMCA_F_IBMCA_ECDSA_SIGN_SIG,
+                     IBMCA_R_EC_INTERNAL_ERROR);
+            return NULL;
+        }
+
+        kinv = NULL;
+        r = NULL;
+        ctx = BN_CTX_new();
+        sign_setup_sw(eckey, ctx, &kinv, &r);
+        BN_CTX_free(ctx);
+        sig = sign_sig_sw(dgst, dgst_len, kinv, r, eckey);
+        goto end2;
+    }
+
+    /* Construct ECDSA_SIG object from char array */
+    r = BN_bin2bn(sigret, privlen, NULL);
+    s = BN_bin2bn(sigret + privlen, privlen, NULL);
+    sig = ECDSA_SIG_new();
+
+end:
+ #ifndef OLDER_OPENSSL
+    if (sig)
+        ECDSA_SIG_set0(sig, r, s);
+ #else
+    if (sig) {
+        BN_free(sig->r);
+        sig->r = r;
+        BN_free(sig->s);
+        sig->s = s;
+    }
+ #endif
+
```

Comment:
The freeing of r and s looks strange.

Reply:
its correct,  ECDSA_SIG_new also allocates structure members in 1.0.2